File: SWEA3316. 동아리실 관리하기/백동승.py
def solve(idx):
    for i in range(1,16):
        for j in range(1,16):
            if(i & j) and (i & dic[s[idx]]):
                dp[idx+1][i] %= 1000000007
                dp[idx+1][i] += dp[idx][j]
 
 
T = int(input())
for test_case in range(1, T + 1):
    s = input()
    s_len = len(s)
    dic = {"A":0b0001,"B":0b0010,"C":0b0100,"D":0b1000}
    dp = [[0 for j in range(16)] for i in range(s_len+1)]
    dp[0][1] = 1;
    for i in range(s_len):
        solve(i)
    ret = 0
    for i in range(1,16):
        ret += dp[s_len][i]
    print("#{} {}".format(test_case, ret%1000000007))

File: SWEA3316. 동아리실 관리하기/백동승.java
import java.io.*;
 
 
class Solution
{
    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static long[][] dp;
    private static String s;
    private static void solve(int idx){
        for(int i = 1; i < 16; i++)
            for(int j = 1; j < 16; j++)
                if((i&j) != 0 && (i & (1 << (s.charAt(idx)-'A'))) != 0){
                    dp[idx+1][i] %= 1000000007;
                    dp[idx+1][i] += dp[idx][j];
                }
    }
     
    public static void main(String args[]) throws Exception
    {
 
        int T = Integer.parseInt(br.readLine().trim());
 
 
        for(int test_case = 1; test_case <= T; test_case++)
        {
            s = br.readLine();
            dp = new long[s.length()+1][];
            for(int i = 0; i < s.length()+1; i++) dp[i] = new long[16];
            dp[0][1] = 1;
            for(int i = 0; i < s.length(); i++) solve(i);
            long ret = 0;
            for(int i = 1; i < 16; i++) ret += dp[s.length()][i];
            bw.write("#"+test_case+" "+ret%1000000007 + "\n");
 
        }
        bw.flush();
        br.close();
        bw.close();
    }
}
